engine/src/main/java/org/hibernate/validator/internal/properties/Signature.java

```diff
@@ -0,0 +1,51 @@
+/*
+ * Hibernate Validator, declare and validate application constraints
+ *
+ * License: Apache License, Version 2.0
+ * See the license.txt file in the root directory or <http://www.apache.org/licenses/LICENSE-2.0>.
+ */
+package org.hibernate.validator.internal.properties;
+
+import java.util.Arrays;
+
+public class Signature {
+
+	private String name;
```

Comment:
I think the two main fields need to be `final` (even if there's a cost): as these keys are getting entered in concurrent maps I'd not try to play smart games with this.

The "hashCode" is different; it's a bit special as it's not a reference so you _could_ avoid the final qualifier, but since you're already writing the other two fields as final you're already going to flush all state so I suppose there's no benefit in keeping that lazy.

I'm not sure if it's worth using that field though? Are you sure the hashcode is always going to be used? There's also a tradeoff with it consuming a precious field space.

Reply:
BTW just checked with jol: that `int` field is costing you 8 bytes because of alignemnt. I'd say it's not worth it.

Reply:
Yes, it's going to be used and it has been proven very useful in HV in general.

Reply:
OK for making the fields final.

Reply:
Regarding the `int`, I don't have data to make a strong suggestion but consider one thing:

in this case you're trading off a little CPU cost for 8 bytes. The 8 bytes are "free" in the context of a micro-benchmark, but they are not in a more complex real-world benchmark.

So maybe you're right and it's the better choice, but don't fool yourself in thinking that JMH is giving you this certainty.

Reply:
OK, I've done the changes, will run a benchmark later today.

Reply:
Benchmarks are not better but not worse either so let's get rid of the field.